### ISSUE TITLE
StochasticTensor strange behavior

### ISSUE THREAD

**Issue by denmoroz** [tensorflow/tensorflow#13473]
I've noticed strange behavior of StochasticTensor. Please, look at this peace of code: 

    inputs = tf.placeholder(shape=(1, 10, ), name="inputs", dtype=tf.float32)

    outputs = tf.layers.dense(inputs, units=10, use_bias=False, activation=tf.nn.sigmoid)
    outputs = st.StochasticTensor(distributions.Bernoulli(probs=outputs, dtype=tf.int32))
    outputs = tf.reshape(outputs, shape=(-1, ))
    
    init_op = tf.group(
        tf.global_variables_initializer(),
        tf.local_variables_initializer()
    )
    
    with tf.Session() as sess:
        sess.run(init_op)
        x = np.random.rand(1, 10)
        
        tf.set_random_seed(2017)
        z1 = sess.run(outputs, feed_dict={inputs: x})
        
        tf.set_random_seed(2017)
        z2 = sess.run(outputs, feed_dict={inputs: x})
    
    print(z1)
    print(z2)

As the result I get:
`
[1 0 1 0 0 1 1 1 0 0]
[0 1 0 1 0 1 0 1 1 1]
`

But numpy sampling has another behavior:

        np.random.seed(2017)
        x = np.random.randint(0, 10, size=10)

        np.random.seed(2017)
        y = np.random.randint(0, 10, size=10)

        print(x)
        print(y)

And (again) the result:
`
[9 6 8 2 3 7 8 0 8 6]
[9 6 8 2 3 7 8 0 8 6]
`

Is it an issue?

P.S.
Tensorflow v1.3.0-rc2-20-g0787eee

**Comment by yaroslavvb** [comment]
Nope, not a bug, you are using Bernoulli distribution in your code which is over 0/1. (ps these kinds of questions are better for stackoverflow)

**Comment by denmoroz** [comment]
I understand :) The problem is that tf.set_random_seed do not affect "randomness". In numpy we have another behavior.

    tf.set_random_seed(2017)
    z1 = sess.run(outputs, feed_dict={inputs: x})
    
    tf.set_random_seed(2017)
    z2 = sess.run(outputs, feed_dict={inputs: x})

So, it looks like a bug :)

**Comment by yaroslavvb** [comment]
cc @ebrevdo  -- how are you supposed to fix the seed of stochastic tensors?

Here's a self-contained example:

```
import os
os.environ['CUDA_VISIBLE_DEVICES']=''
import tensorflow as tf
import numpy as np
from tensorflow.contrib.bayesflow.python.ops import stochastic_tensor_impl
from tensorflow.python.ops import distributions
st = stochastic_tensor_impl

tf.reset_default_graph()

outputs = st.StochasticTensor(distributions.bernoulli.Bernoulli(probs= [[0.5]*10], dtype=tf.int32))
outputs = tf.reshape(outputs, shape=(-1, ))

sess = tf.Session()
sess.run(tf.global_variables_initializer())
    
for i in range(10):
  tf.set_random_seed(2017)
  print(sess.run(outputs))
```

**Comment by ebrevdo** [comment]
You'll get the same behavior even if you set a fixed seed:

a = tf.random_uniform([3], seed=5)

 for i in range(4):
    tf.set_random_seed(2017)
    print(s.run(a))
   ...:
[ 0.045017    0.30386245  0.43822038]
[ 0.04852033  0.07889724  0.8716917 ]
[ 0.98431134  0.39827144  0.79930592]
[ 0.92832315  0.6284616   0.49420452]

The behavior that you want would actually require recreating the graph:

for i in range(4):
    tf.reset_default_graph()
    tf.set_random_seed(2017)
    a = tf.random_uniform([3], seed=5)
    print(tf.Session().run(a))

[ 0.045017    0.30386245  0.43822038]
[ 0.045017    0.30386245  0.43822038]
[ 0.045017    0.30386245  0.43822038]
[ 0.045017    0.30386245  0.43822038]

and while it's true there's currently no way to set a seed for a
StochasticTensor, that's not the crux of the problem here.  See the
documentation here
<https://www.tensorflow.org/api_docs/python/tf/set_random_seed>.

**Comment by yaroslavvb** [comment]
Ah right, the seed complexity bites again.

@denmoroz There are essentially three things affecting randomness -- graph level seed, operation level seed, and operation current state. Setting seeds will guarantee the same pseudo-random sequence of numbers for an operation. If you want to generate a sequence of constants you need to recreate the operation.

This is simpler in numpy because np.random calls shared core, rather than creating operation with persistent state

**Comment by denmoroz** [comment]
@yaroslavvb Thanks! Yeah, it is not easy to understand such behavior cause it differs from the previous experience. Seems like tf.set_random_seed should be renamed :)

**Comment by ebrevdo** [comment]
If you want to always control your randomness exactly, check out this module
<https://www.tensorflow.org/api_docs/python/tf/contrib/stateless>.

**Comment by reedwm** [comment]
Closing this issue because an @ebrevdo provided an explanation of the seed behavior.
